engine/sync_manager.go

```diff
@@ -397,72 +361,67 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
-				if err != nil {
-					m.currencyPairs[x].Ticker.NumErrors++
-				}
-				m.currencyPairs[x].Ticker.HaveData = true
-				m.currencyPairs[x].Ticker.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s ticker sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			case SyncItemOrderbook:
-				origHadData := m.currencyPairs[x].Orderbook.HaveData
-				m.currencyPairs[x].Orderbook.LastUpdated = time.Now()
-				if err != nil {
-					m.currencyPairs[x].Orderbook.NumErrors++
-				}
-				m.currencyPairs[x].Orderbook.HaveData = true
-				m.currencyPairs[x].Orderbook.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s orderbook sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			case SyncItemTrade:
-				origHadData := m.currencyPairs[x].Trade.HaveData
-				m.currencyPairs[x].Trade.LastUpdated = time.Now()
-				if err != nil {
-					m.currencyPairs[x].Trade.NumErrors++
-				}
-				m.currencyPairs[x].Trade.HaveData = true
-				m.currencyPairs[x].Trade.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s trade sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			}
-		}
+	c, exists := m.currencyPairs[k]
+	if !exists {
+		return fmt.Errorf("%w for %s %s %s %s", errCouldNotSyncNewData, k.Exchange, k.Pair, k.AssetType, syncType)
+	}
+
+	c.locks[syncType].Lock()
+	defer c.locks[syncType].Unlock()
+
+	if c.trackers[syncType] == nil {
+		c.trackers[syncType] = &syncBase{}
+	}
+	s := c.trackers[syncType]
+
+	if !s.IsUsingWebsocket {
+		s.IsUsingWebsocket = true
+		s.IsUsingREST = false
+		log.Warnf(log.SyncMgr,
+			"%s %s %s: %s Websocket re-enabled, switching from rest to websocket",
+			c.Exchange,
+			m.FormatCurrency(c.Pair),
+			strings.ToUpper(c.AssetType.String()),
+			syncType,
+		)
 	}
-	return fmt.Errorf("%w for %s %s %s", errCouldNotSyncNewData, exchangeName, p, a)
+
+	return m.update(c, syncType, err)
+}
+
+// update notifies the syncManager to change the last updated time for a exchange asset pair
+func (m *syncManager) update(c *currencyPairSyncAgent, syncType syncItemType, err error) error {
+	if syncType < SyncItemTicker || syncType > SyncItemTrade {
+		return fmt.Errorf("%v %w", syncType, errUnknownSyncItem)
+	}
+
+	s := c.trackers[syncType]
+
+	origHadData := s.HaveData
+	s.LastUpdated = time.Now()
+	if err != nil {
+		s.NumErrors++
+	}
+	s.HaveData = true
+	s.IsProcessing = false
```

Comment:
You mention in the PR description that you've fixed a race condition with this. Its true, but this is also the only reference to the field now, and the bool defaults to false anyway. Seems like you could remove it?

Reply:
Agreed. Done. ✅